ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,526 @@
+   * @param messages a list of messages without content
+   * @returns either a list of message with content or a query error
+   */
+  const getAllMessagesWithContent = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<MessageWithContent>
+  > => {
+    if (messages.length) {
+      // this spread is needed as typescript wouldn't recognize messages[0] to be defined otherwise
+      const [firstMessage, ...otherMessages] = messages;
+      return sequenceT(taskEither)(
+        getMessageWithContent(firstMessage),
+        ...otherMessages.map(getMessageWithContent)
+      );
+    }
+    return taskEither.of([]);
+  };
+
+  /**
+   * Given a list of messages, it queires for relative notifications
+   * @param messages
+   */
+  const findNotificationsForAllMessages = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<RetrievedNotification>
+  > => {
+    if (messages.length) {
+      // this spread is needed as typescript wouldn't recognize messages[0] to be defined otherwise
+      const [firstMessage, ...otherMessages] = messages;
+
+      return sequenceT(taskEither)(
+        fromQueryEither<ReadonlyArray<RetrievedNotification>>(
+          () =>
+            iteratorToArray(
+              notificationModel.findNotificationsForMessage(firstMessage.id)
+            ),
+          "findNotificationsForRecipient"
+        ),
+        ...otherMessages.map(m =>
+          fromQueryEither<ReadonlyArray<RetrievedNotification>>(
+            () =>
+              iteratorToArray(
+                notificationModel.findNotificationsForMessage(m.id)
+              ),
+            "findNotificationsForRecipient"
+          )
+        )
+      ).foldTaskEither(
+        e => fromEither(left(e)),
+        arrayOfArray =>
+          fromEither(
+            right(
+              arrayOfArray.reduce(
+                /* flatten */ (flat, elem) => [...flat, ...elem],
+                []
+              )
+            )
+          )
+      );
+    }
+    return taskEither.of([]);
+  };
+
+  const findAllNotificationStatuses = (
+    notifications: ReadonlyArray<RetrievedNotification>
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<NotificationStatus>
+  > => {
+    if (notifications.length) {
+      // this spread is needed as typescript wouldn't recognize queries[0] to be defined otherwise
+      const [firstQuery, ...otherQueries] = notifications
+        .reduce(
+          (q, n) => [
+            ...q,
+            [n.id, NotificationChannelEnum.EMAIL],
+            [n.id, NotificationChannelEnum.WEBHOOK]
+          ],
+          []
+        )
+        .map(([notificationId, channel]) =>
+          fromQueryEither(
+            () =>
+              notificationStatusModel.findOneNotificationStatusByNotificationChannel(
```

Comment:
why do we need to differentiate by channel?

Reply:
no, but I wanted to use the actual method which queries by statusId, and calculates the statusId by its channel

Reply:
no since they're not needed to rebuild the whole thing